### PR TITLE
[MIR] Refactor the MIR translator to use LLVM Builder directly

### LINKED ISSUE #31003: [MIR] Should translator use LLVM builder more directly?
Currently we use various `build::*` functions in order to build up the function, but many of the checks done by the `build::*` functions are not necessary for us.

Namely, those functions are concerned with checking for termination and unreachability. Both of these concerns are not relevant to MIR – we have well specified terminator system and unreachable blocks are removed before MIR gets to the trans.

Suggestion is to refactor the MIR translator to use the `trans::builder::Builder` directly.

### PR DESCRIPTION
**Pull request by pczarn**
Closes #31003

### CODE REVIEW

File: src/librustc_trans/trans/common.rs [hunk 1]
```diff
@@ -656,6 +657,14 @@ impl<'blk, 'tcx> BlockS<'blk, 'tcx> {
                                          self.fcx.param_substs,
                                          value)
     }
+
+    pub fn build(&self) -> Builder<'blk, 'tcx> {
+        // Create a fresh builder from the crate context.
+        let builder = self.ccx().builder();
+        // Set the builder's position to this block's end.
+        builder.position_at_end(self.llbb);
```
Comment (nagisa) [inline src/librustc_trans/trans/common.rs]: One of the points (other than getting rid of the unnecessary checks) was to get rid of repositioning the builder for every instruction. This should be not necessary (not in MIR, at least). I think it would be better to have something like

```
struct MIRBlock<...> {
     builder: Builder<...>
     bcx: Block<...>
}

// split out the `Builder` trait implemented by the `Builder` struct in trans/builder, and
impl<...> Builder for MIRBlock<...> {
    // has all the functions that produce LLVM insns (likely generated by a macro or something)
}
```

so you could just call

```
bcx.br(self.llblock(target))
```

anywhere in the MIR, instead of doing the `B(bcx).br(...)` or `bcx.builder().br()` dance every time we emit an instruction.

File: src/librustc_trans/trans/mir/block.rs [hunk 2]
```diff
@@ -39,16 +39,17 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
 
         debug!("trans_block: terminator: {:?}", data.terminator());
 
+        // FIXME debuginfo: apply DebugLoc before all uses of `build()`.
```
Comment (nagisa) [inline src/librustc_trans/trans/mir/block.rs]: We currently do not emit debug info anywhere in the MIR and there’s a explicit issue for that, so this FIXME is not necessary. Feel free to also remove all the `DebugLoc::None`s you find around as part of this refactoring.

File: src/librustc_trans/trans/common.rs (hunk 1 above)
Comment (pczarn) [inline src/librustc_trans/trans/common.rs]: Each crate has only one builder, which `BuilderRef` points to. I'm worried the lack of repositioning could introduce bugs. For example, after the call to `MirContext.unreachable_block`, which may emit an instruction, the builder for bcx must be repositioned (I think).

Also, keeping builder refs alongside blocks in the table would add at least several bytes of overhead for each block. Otherwise, you'd need a different dance every time bcx is passed from MIR to common trans.

Comment (nagisa) [inline src/librustc_trans/trans/common.rs]: > Each crate has only one builder, which BuilderRef points to.

You can create LLVM builders at will. The `BuilderRef` in `LocalCrateContext` is just a cached builder for the purposes of the old translator. No need to restrict ourselves only to the things available to the old translator.

> I'm worried the lack of repositioning could introduce bugs.

True, especially when there could be two different builders acting on the same basic block (i.e. bcx is passed to some `build::*` function), but the bugs are easily preventable by making `bcx` in `BlockBuilder` private and adding a method similar to this:

```rust
fn with_old_bcx<R, F: FnOnce(b: Block<...>) -> R> (&self, f: F) {
    f(self.bcx);
    self.builder.position_at_end(self.bcx.llbb);
}
```

> Also, keeping builder refs alongside blocks in the table would add at least several bytes of overhead for each block.

I don’t find that to be a problem. At any given point the translator is dealing with a small amount of blocks. I don’t think there’s any place in MIR trans where it would be necessary to have more than 2 or 3 `BlockBuilder`s alive at any given time. If you are worried about allocation overhead, dropped builders could be reused in some way as well.

> Otherwise, you'd need a different dance every time bcx is passed from MIR to common trans.

`bulder.with_old_bcx(|bcx| some_old_trans_function_wants_bcx(bcx, ...))` is much more trivial than sprinkling a `bcx.builder()` everywhere, where MIR trans wants to emit an instruction. Consider, that the old trans will be phased out in the future and we should adapt (once the old trans is gone) the common trans utilities to match the MIR trans model, and not the other way around.

All in all, while this PR indeed makes MIR’s use of LLVM builder more directly, in my opinion there’s much larger refactoring opportunities available. If you are not interested in exploring them, I’ll gladly r+ this if you change all `bcx.build()` to be `B(bcx)` (where `B` == `build::B`), because it does essentially the same thing as this method, but is shorter to type out.

File: src/librustc_trans/trans/common.rs [hunk 3]
```diff
@@ -659,8 +661,119 @@ impl<'blk, 'tcx> BlockS<'blk, 'tcx> {
...
+    }
+
+    pub fn monomorphize<T>(&self, value: &T) -> T
+        where T: TypeFoldable<'tcx>
+    {
+        self.bcx.monomorphize(value)
+    }
+}
+
+impl<'blk, 'tcx> Deref for BlockAndBuilder<'blk, 'tcx> {
```
Comment (nagisa) [inline src/librustc_trans/trans/common.rs]: Nice!

File: src/librustc_trans/trans/mir/block.rs [hunk 4]
```diff
@@ -30,7 +30,7 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
     pub fn trans_block(&mut self, bb: mir::BasicBlock) {
         debug!("trans_block({:?})", bb);
 
-        let mut bcx = self.bcx(bb);
+        let mut bcx = self.bcx(bb).build();
```
Comment (nagisa) [inline src/librustc_trans/trans/mir/block.rs]: It might be better to just return a `BlockAndBuilder` from `self.bcx(BasicBlock)`.

File: src/librustc_trans/trans/mir/block.rs [hunk 5]
```diff
@@ -254,44 +260,46 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
...
-            None => {
-                let bl = self.fcx.new_block("unreachable", None);
-                build::Unreachable(bl);
-                self.unreachable_block = Some(bl);
-                bl
-            }
-        }
+        self.unreachable_block.unwrap_or_else(|| {
+            let bl = self.fcx.new_block("unreachable", None);
+            build::B(bl).unreachable();
```
Comment (nagisa) [inline src/librustc_trans/trans/mir/block.rs]: Why not `bl.build()`? Sure, the `build::B()` will not create a whole new builder (a temporary downside), but I think consistency should be priority.

**Comment by nagisa** [comment]
I really like what this is looking like. I do not find lack of builder recycling to be a big problem or a blocker to land this.

I’ll let this stay around for a day or two so @nikomatsakis gets a chance to see it, but otherwise r=me with the two nits (and the tidy failure) fixed.

**Comment by pczarn** [comment]
Finished. I have some thoughts: almost all methods in the `mir` directory could be implemented for `Block` rather than `MirContext`.

Storing `LandingPad`s directly in blocks made them bigger. I changed that. When MIR finally replaces trans, we should check how much we can save by making blocks smaller.

File: src/librustc_trans/trans/base.rs [hunk 6]
```diff
@@ -1616,6 +1616,7 @@ pub fn new_fn_ctxt<'a, 'tcx>(ccx: &'a CrateContext<'a, 'tcx>,
         param_substs: param_substs,
         span: sp,
         block_arena: block_arena,
+        lpad_arena: TypedArena::new(),
```
Comment (nagisa) [inline src/librustc_trans/trans/base.rs]: You might want to leave off this change for a separate PR, honestly (or at least factor it out into a separate commit, so it easier to see related changes).

Code related to this field is very precondition-sensitive, I’m worried changing it like this might invalidate some of those preconditions. E.g. MSVC landing blocks, especially, are on per-block basis and it is likely to be wrong there.

File: src/librustc_trans/trans/mir/block.rs [hunk 7]
```diff
@@ -254,48 +269,54 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
...
         } else {
             let llretty = Type::struct_(ccx, &[Type::i8p(ccx), Type::i32(ccx)], false);
-            let slot = base::alloca(bcx, llretty, "personalityslot");
+            let slot = bcx.with_block(|bcx| {
+                base::alloca(bcx, llretty, "personalityslot")
+            });
             self.llpersonalityslot = Some(slot);
-            base::call_lifetime_start(bcx, slot);
+            bcx.with_block(|bcx| {
+                base::call_lifetime_start(bcx, slot);
```
Comment (nagisa) [inline src/librustc_trans/trans/mir/block.rs]: You can move this into the `with_block` above.

File: src/librustc_trans/trans/base.rs (hunk 6 above)
Comment (nagisa) [inline src/librustc_trans/trans/base.rs]: Ah, I see it, you’re still storing a pointer inside a `Block`. A few thoughts:

1. Size of `BlockS` does not really matter, because we only ever pass a `Block` around, which is a `pub type Block<'blk, 'tcx> = &'blk BlockS<'blk, 'tcx>;` – i.e. is a reference and there’s barely any performance overhead passing a pointer to larger or smaller struct;
2. From my experimentation with implementing MSVC LPs in MIR, all that really needs to be stored around for MIR itself is a single ValueRef (a pointer, basically) inside MirContext. All the landing-pad related things in `Block` are for the old trans to use, and I wouldn’t bother much changing it (but since you’ve already changed it, its fine, I guess?).

I’d still like it split up into a separate commit so it is easier to review and these changes do not mix with the Builder refactor.

File: src/librustc_trans/trans/cleanup.rs [hunk 8]
```diff
@@ -1039,15 +1039,15 @@ impl EarlyExitLabel {
     ///
     /// Returns a new label which will can be used to cache `bcx` in the list of
     /// early exits.
-    fn start(&self, bcx: Block) -> EarlyExitLabel {
+    fn start(&self, mut bcx: Block) -> EarlyExitLabel {
```
Comment (nagisa) [inline src/librustc_trans/trans/cleanup.rs]: Causes a travis failure.

File: src/librustc_trans/trans/common.rs [hunk 9]
```diff
@@ -429,14 +434,19 @@ impl<'a, 'tcx> FunctionContext<'a, 'tcx> {
 
     pub fn new_block(&'a self,
                      name: &str,
-                     opt_node_id: Option<ast::NodeId>)
+                     opt_node_id: Option<ast::NodeId>,
+                     landing_pad: Option<LandingPad>)
```
Comment (nagisa) [inline src/librustc_trans/trans/common.rs]: Most of `new_block` calls pass in `None` for `landing_pad`, it seems. If you’re adding this argument just for MIR, you shouldn’t bother, because the current code which sets the field is just a placeholder and I don’t believe we will be setting this field in MIR on block construction once SEH EH gets implemented (it is quite likely we won’t be setting it at all, even).

File: src/librustc_trans/trans/mir/block.rs [hunk 10]
```diff
@@ -143,39 +149,41 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
...
-                                      unreachable_blk.llbb,
-                                      landingpad.llbb,
-                                      Some(attrs),
-                                      debugloc);
+                        let bundle = bcx.lpad().and_then(|b| b.bundle());
+                        bcx.invoke(callee.immediate(),
+                                   &llargs[..],
+                                   unreachable_blk.llbb,
+                                   landingpad.llbb(),
+                                   bundle,
```
Comment (nagisa) [inline src/librustc_trans/trans/mir/block.rs]: Just pass `None` please. This code will most certainly change with SEH EH implementation.

File: src/librustc_trans/trans/mir/block.rs [hunk 11]
```diff
@@ -254,48 +269,54 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
...
-        // FIXME(#30941) this doesn't handle msvc-style exceptions
-        *bcx.lpad.borrow_mut() = Some(LandingPad::gnu());
+    fn make_landing_pad(&mut self,
+                        cleanup: BlockAndBuilder<'bcx, 'tcx>)
+                        -> BlockAndBuilder<'bcx, 'tcx>
+    {
+        let cleanup_llbb = cleanup.llbb();
+        let bcx = cleanup.map_block(|cleanup| {
+            // FIXME(#30941) this doesn't handle msvc-style exceptions
+            cleanup.fcx.new_block("cleanup", None, Some(LandingPad::gnu()))
```
Comment (nagisa) [inline src/librustc_trans/trans/mir/block.rs]: Preexisting, but `fcx` is available on `self` AFAIR, so you can just do

```
self.fcx.new_block("cleanup", None)
```

**Comment by nagisa** [comment]
> almost all methods in the mir directory could be implemented for Block rather than MirContext

It doesn’t make much sense, does it? All the methods operate strictly on MIR, don’t they. It certainly might be a case of `mir` directory contents getting lifted up into `trans` over time, though.

File: src/librustc_trans/trans/mir/operand.rs [hunk 12]
```diff
@@ -164,11 +169,21 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
...
                          lldest: ValueRef,
                          operand: OperandRef<'tcx>)
     {
         debug!("store_operand: operand={}", operand.repr(bcx));
+        bcx.with_block(|bcx| {
+            self.store_operand_direct(bcx, lldest, operand)
+        })
+    }
+
+    pub fn store_operand_direct(&mut self,
```
Comment (nagisa) [inline src/librustc_trans/trans/mir/operand.rs]: Why not inline?

File: src/librustc_trans/trans/mir/operand.rs [hunk 13]
```diff
@@ -164,11 +169,21 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
...
 
     pub fn store_operand(&mut self,
-                         bcx: Block<'bcx, 'tcx>,
+                         bcx: &BlockAndBuilder<'bcx, 'tcx>,
                          lldest: ValueRef,
                          operand: OperandRef<'tcx>)
     {
         debug!("store_operand: operand={}", operand.repr(bcx));
+        bcx.with_block(|bcx| {
+            self.store_operand_direct(bcx, lldest, operand)
```
Comment (nagisa) [inline src/librustc_trans/trans/mir/operand.rs]: Why not inline?

File: src/librustc_trans/trans/mir/rvalue.rs [hunk 14]
```diff
@@ -61,39 +60,43 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
...
-                let base = expr::get_dataptr(bcx, dest.llval);
-                tvec::iter_vec_raw(bcx, base, elem.ty, size, |bcx, llslot, _| {
-                    self.store_operand(bcx, llslot, elem);
-                    bcx
+                let elem = self.trans_operand(&bcx, elem);
+                let size = self.trans_constval(&bcx, &count.value, count.ty).immediate();
+                bcx.map_block(|block| {
+                    let base = expr::get_dataptr(block, dest.llval);
+                    tvec::iter_vec_raw(block, base, elem.ty, size, |block, llslot, _| {
+                        self.store_operand_direct(block, llslot, elem);
```
Comment (pczarn) [inline src/librustc_trans/trans/mir/rvalue.rs]: `fn store_operand_direct` is used here. `iter_vec_raw` is annoying.

**Comment by pczarn** [comment]
Updated.

File: src/librustc_trans/trans/mir/mod.rs [hunk 15]
```diff
@@ -174,17 +170,26 @@ fn arg_value_refs<'bcx, 'tcx>(bcx: Block<'bcx, 'tcx>,
                let lldata = llvm::get_param(fcx.llfn, idx);
                let llextra = llvm::get_param(fcx.llfn, idx + 1);
                idx += 2;
-               let lltemp = base::alloc_ty(bcx, arg_ty, &format!("arg{}", arg_index));
-               build::Store(bcx, lldata, expr::get_dataptr(bcx, lltemp));
-               build::Store(bcx, llextra, expr::get_meta(bcx, lltemp));
+               let lltemp = bcx.with_block(|bcx| {
+                   base::alloc_ty(bcx, arg_ty, &format!("arg{}", arg_index))
+               });
+               let (dataptr, meta) = bcx.with_block(|bcx| {
```
Comment (nagisa) [inline src/librustc_trans/trans/mir/mod.rs]: Merge into `with_block` above and return a `(lltemp, dataptr, meta)`?

File: src/librustc_trans/trans/mir/mod.rs [hunk 16]
```diff
@@ -174,17 +170,26 @@ fn arg_value_refs<'bcx, 'tcx>(bcx: Block<'bcx, 'tcx>,
...
            } else {
                // otherwise, arg is passed by value, so make a
                // temporary and store it there
                let llarg = llvm::get_param(fcx.llfn, idx);
                idx += 1;
-               let lltemp = base::alloc_ty(bcx, arg_ty, &format!("arg{}", arg_index));
-               base::store_ty(bcx, llarg, lltemp, arg_ty);
+               let lltemp = bcx.with_block(|bcx| {
+                   base::alloc_ty(bcx, arg_ty, &format!("arg{}", arg_index))
+               });
```
Comment (nagisa) [inline src/librustc_trans/trans/mir/mod.rs]: Similarly.

**Comment by nagisa** [comment]
Please squash the `f \d` commits into one. r=me once the last two nits are fixed.

**Comment by pczarn** [comment]
Finished.
